### DISCUSSION TITLE
How do I have Vercel install `ffmpeg` before running my server?

### DISCUSSION THREAD

**Discussion by hirehamir** [vercel/vercel#9729] [Help]
### Problem Description

I'm trying to use `pydub` in a Django application and I believe it requires `ffmpeg` installed to use it.

How do I make the Vercel server running my code install `ffmpeg`, say, with a command like `apt-get install -y ffmpeg`, before starting my Django application?

**Comment by wiiiimm** [comment]
While there are ways to get `ffmpeg` to install on Vercel's serverless containers like:

1. using the `@ffmpeg-installer/ffmpeg` npm package,
2. downloading `ffmpeg` as an executable binary (https://vercel.com/docs/concepts/deployments/build-image#installing-additional-packages) and optionally moving it to a cachable location

we generally do not recommend running this kind of server workload as serverless functions and installing `ffmpeg` will very likely push the serverless function over the [50MB limit](https://vercel.com/docs/concepts/limits/overview#serverless-function-size).

**Reply by hirehamir** [reply]
How do I check if Vercel actually ran `"installCommand": "apt install ffmpeg",` correctly?

It seems like it didn't since the development URL still says `[Errno 2] No such file or directory: 'ffprobe'`, which was the same result as when I didn't have an `installCommand` key in `vercel.json`. 

The build logs say

```
Installing required dependencies...
Build Completed in /vercel/output [23s]
```

I don't know if `Installing required dependencies` just refers to the ordinary dependencies for the project, or if that also includes `installCommand`.


`"installCommand": "sudo apt install ffmpeg",` didn't work, either.

This is my complete `vercel.json` file.

```json
{
  "builds": [
    {
      "src": "vercel_app/wsgi.py",
      "use": "@vercel/python"
    }
  ],
  "installCommand": "sudo apt install ffmpeg",
  "routes": [
    {
      "src": "/(.*)",
      "dest": "vercel_app/wsgi.py"
    }
  ]
}
```

**Reply by hirehamir** [reply]
I've got `    "@ffmpeg-installer/ffmpeg": "^1.1.0",` in my `package.json`, but it still seems like Vercel may not be converting an audio file correctly.

1. Is there a way Vercel recommends for converting a file from `.m4a` to `.mp3`, where both the original and converted files are in the `tmp/` directory?
2. How do I `ssh` into my Vercel deployment? I want to do `which ffmpeg` to do some troubleshooting.

**Reply by digitros** [reply]
Hey @hamirmahal, could you solve this? I'm trying to convert audio too.

**Reply by hirehamir** [reply]
@digitros I haven't found a solution, so far.

**Reply by digitros** [reply]
@hamirmahal Yesterday I implemented my usecase with the @ffmpeg-installer/ffmpeg library and it is transforming ogg whatsapp audio files to m4a. No problems so far even with 2-3 minutes of audio. But a few weeks ago I couldn't make it work. The only difference is that I now have the vercel pro plan.

**Reply by hirehamir** [reply]
@digitros Hmm... That's interesting. I just tried again right now, and I see
```
Failed to convert /tmp/fo9sOeRWVu4.m4a to /tmp/fo9sOeRWVu4.mp3: Error: ffmpeg exited with code null and signal null
```
in my logs. I don't have Vercel's professional plan.

This is my JavaScript code.

```js
const ffmpegPath = require('@ffmpeg-installer/ffmpeg').path;
const ffmpeg = require('fluent-ffmpeg');
ffmpeg.setFfmpegPath(ffmpegPath);

/**
 * @param {string} inputFilePath
 * @param {string} outputFilePath
 */
function convertM4AToMP3(inputFilePath, outputFilePath) {
  return new Promise((resolve, reject) => {
    const outputPath = path.parse(outputFilePath).dir;
    const outputFileName = path.parse(outputFilePath).name;

    ffmpeg(inputFilePath)
      .outputOptions(['-vn', '-acodec', 'libmp3lame', '-qscale:a', '2'])
      .output(`${outputPath}/${outputFileName}.mp3`)
      .on('end', () => {
        console.log(
          `Successfully converted ${inputFilePath} to ${outputFilePath}`
        );
        resolve();
      })
      .on('error', (err) => {
        console.error(
          `Failed to convert ${inputFilePath} to ${outputFilePath}: ${err}`
        );
        reject(err);
      })
      .run();
  });
}
```

Would you mind sharing the code that lets you convert audio files without issue on a deployed Vercel server?
